ee/localserver/krypto-ec-middleware.go

```diff
@@ -459,3 +477,197 @@ func (bhr *bufferedHttpResponse) WriteHeader(code int) {
 func (bhr *bufferedHttpResponse) Bytes() []byte {
 	return bhr.buf.Bytes()
 }
+
+func (e *kryptoEcMiddleware) presenceDetectionCallback(challengeBox *challenge.OuterChallenge) {
+	// extract cmd req from challenge box
+	ctx, cancel := context.WithTimeout(context.Background(), presencedetection.DetectionTimeout)
+	defer cancel()
+
+	var cmdReq v2CmdRequestType
+	if err := json.Unmarshal(challengeBox.RequestData(), &cmdReq); err != nil {
+		e.slogger.ErrorContext(ctx,
+			"error unmarshaling cmd request",
+			"err", err,
+		)
+		return
+	}
+
+	// figure out if we need to do presence detection
+	detectionIntervalStr, ok := cmdReq.Headers[kolidePresenceDetectionIntervalHeaderKey]
+	if !ok || len(detectionIntervalStr) == 0 {
+		return
+	}
+
+	// no presence detection requested
+	if len(detectionIntervalStr) == 0 || detectionIntervalStr[0] == "" {
+		return
+	}
+
+	// presence detection is not yet available on linux
+	if runtime.GOOS == "linux" {
+		return
+	}
+
+	detectionIntervalDuration, err := time.ParseDuration(detectionIntervalStr[0])
+	if err != nil {
+		return
+	}
+
+	reasonKey := kolidePresenceDetectionReasonMacosHeaderKey
+	if runtime.GOOS == "windows" {
+		reasonKey = kolidePresenceDetectionReasonWindowsHeaderKey
+	}
+
+	reason := "authenticate"
+	if reasonStr, ok := cmdReq.CallbackHeaders[reasonKey]; ok && len(reasonStr) > 0 && len(reasonStr[0]) > 0 {
+		reason = reasonStr[0]
+	}
+
+	if !e.presenceDetectionLock.TryLock() {
+		e.slogger.Log(ctx, slog.LevelInfo,
+			"dropping presence detection callback, already in progress",
+		)
+		return
+	}
+	defer e.presenceDetectionLock.Unlock()
+
+	type detectionResult struct {
+		durtionSinceLastDetection time.Duration
+		err                       error
+	}
+
+	detectionDoneChan := make(chan *detectionResult)
+
+	// kick of presence detection
+	gowrapper.Go(ctx, e.slogger, func() {
+		durationSinceLastDetection, err := e.presenceDetector.DetectPresence(reason, detectionIntervalDuration)
+		if err != nil {
+			e.slogger.Log(ctx, slog.LevelInfo,
+				"presence_detection",
+				"reason", reason,
+				"interval", detectionIntervalDuration,
+				"duration_since_last_detection", durationSinceLastDetection,
+				"err", err,
+			)
+		}
+
+		detectionDoneChan <- &detectionResult{
+			durtionSinceLastDetection: durationSinceLastDetection,
+			err:                       err,
+		}
+	})
+
+	callbackTicker := time.NewTicker(e.presenceDetectionCallbackInterval)
+	defer callbackTicker.Stop()
+
+	presenceDetectionStartTime := time.Now()
+
+	var returnedDetectionResult *detectionResult
+
+	headers := map[string][]string{
+		kolideArchHeaderKey: {runtime.GOARCH},
+		kolideOsHeaderKey:   {runtime.GOOS},
+	}
+
+	kolideSessionId, ok := cmdReq.CallbackHeaders[kolideSessionIdHeaderKey]
+	if ok {
+		headers[kolideSessionIdHeaderKey] = kolideSessionId
+	}
+
+	for {
```

Comment:
I think a comment at the top here would help -- `Send status update callbacks to k2 on an interval, waiting until presence detection completes` or something

Reply:
done